### ISSUE TITLE
[BUG]Engine Image fail to reach `ready` state if there are tainted worker node

### ISSUE THREAD

**Issue by yasker** [longhorn/longhorn#2081]
**Describe the bug**
Engine Image fail to reach `ready` state if there are tainted worker node in the cluster and Longhorn doesn't tolerate the taint.

**To Reproduce**
1. Install Longhorn
2. Taint one of the node, then delete the engine image daemonset to allow recreation
3. Volume operations (attach/detach) will fail due to engine image is not ready.

**Expected behavior**
Volume operations should still work.

**Log**
```
2020-12-09T13:34:19.849639616-05:00 time="2020-12-09T18:34:19Z" level=error msg="Error in request: unable to attach volume pvc-xxx to xxx: cannot attach volume pvc-xxx7 with image longhornio/longhorn-engine:v1.0.2: engine image ei-ee18f965 (longhornio/longhorn-engine:v1.0.2) is not ready, it's deploying"
```
```
typemeta:
    kind: ""
    apiversion: ""
  objectmeta:
    name: engine-image-ei-ee18f965
...
  status:
    currentnumberscheduled: 6
    numbermisscheduled: 1
    desirednumberscheduled: 6
    numberready: 6
    observedgeneration: 1
    updatednumberscheduled: 6
    numberavailable: 6
    numberunavailable: 0
    collisioncount: null
    conditions: []
```

**Environment:**
 - Longhorn version: v1.0.2
 - Kubernetes distro (e.g. RKE/K3s/EKS/OpenShift) and version: k3s v1.18.6
 - Node config
   - OS type and version: RHEL
   - CPU per node: 4
   - Memory per node: 8
   - Disk type(e.g. SSD/NVMe): SSD
   - Network bandwidth and latency between the nodes: 10GB
 - Underlying Infrastructure (e.g. on AWS/GCE, EKS/GKE, VMWare/KVM, Baremetal): VMWare

**Additional context**
It's due to the tainted node prevent the daemonset to reach the full availability, result in engine image object failed to reach ready state.

Ideally, each manager on the node should only care about the engine image status of it's node rather than the whole readiness or not.

**Comment by PhanLe1010** [comment]
## Proposal

**Goal**: we want to change from checking EI's availability on all nodes to checking the EI's availability on some nodes (the attaching node and all replicas' nodes)

### Changes to EI's `EngineImageStatus` struct

1. Replace [EngineImageStateReady](https://github.com/longhorn/longhorn-manager/blob/c2e1728c2dbfb05f8060bc182c9ad5c23d4171ee/types/resource.go#L255) by `EngineImageStateDeployed` which indicate that the EI is fully deployed on all nodes
2. Add `NodeDeploymentMap map[string]bool` into [EngineImageStatus](https://github.com/longhorn/longhorn-manager/blob/c2e1728c2dbfb05f8060bc182c9ad5c23d4171ee/types/resource.go#L271). This map keeps track of the list of nodes that the EI DaemonSet successfully deployed EI to. 
3. Inside [syncEngineImage()](https://github.com/longhorn/longhorn-manager/blob/c2e1728c2dbfb05f8060bc182c9ad5c23d4171ee/controller/engine_image_controller.go#L182), we watch the pods of EI DaemonSet and update the `EngineImageStateReady` and NodeDeploymentMap

### Modify EI readiness checking condition in the below places

|  Location  |  Checking Condition |  
|---|---|
| [GetEngineClient()](https://github.com/longhorn/longhorn-manager/blob/c2e1728c2dbfb05f8060bc182c9ad5c23d4171ee/manager/engine.go#L203)   |  Only need to check if the EI's `Status.NodeDeploymentMap` has the current node ID set to true| 
| [Create() volume](https://github.com/longhorn/longhorn-manager/blob/c2e1728c2dbfb05f8060bc182c9ad5c23d4171ee/manager/volume.go#L214)| Don’t need to check EI readiness |
| [Attach()](https://github.com/longhorn/longhorn-manager/blob/c2e1728c2dbfb05f8060bc182c9ad5c23d4171ee/manager/volume.go#L285) API  | Need to check EI on the attaching node as well as all replicas’ nodes  of the volume |
| [EngineUpgrade API](https://github.com/longhorn/longhorn-manager/blob/c2e1728c2dbfb05f8060bc182c9ad5c23d4171ee/manager/volume.go#L681)  |  Need to check EI on the attaching node as well as all replicas’ node  of the volume |
| Auto upgrade engine | Need to check EI on the attaching node as well as all replicas’ node  of the volume |
| [upgradeEngineForVolume](https://github.com/longhorn/longhorn-manager/blob/c2e1728c2dbfb05f8060bc182c9ad5c23d4171ee/controller/volume_controller.go#L1541) |  Check EI on the attaching node as well as all replicas’ node of the volume|
| Replica Scheduling | Only schedule replica onto nodes that have the corresponding EI |

**Comment by PhanLe1010** [comment]
## Manual Tests:
### Case 1: Test volume operations when engine image DaemonSet is miss scheduled 
1. Install Longhorn in a 3-node cluster: `node-1`, `node-2`, `node-3`
1. Create a volume, `vol-1`, of 3 replicas
1. 1. Create another volume, `vol-2`, of 3 replicas
1. Taint `node-1` with the taint: `key=value:NoSchedule`
1. Check that all functions (attach, detach, snapshot, backup, expand, restore, ... ) are working ok for `vol-1`

### Case 2: Test volume operations engine image DaemonSet is not fully deployed
1. Continue from case 1
1. Attach `vol-1` to `node-1`. Change the number of replicas of `vol-1` to 2. Delete the replica on `node-1`
1. Delete the pod on `node-1` of the engine image DaemonSet. Or delete the engine image DaemonSet and wait for Longhorn to automatically recreates it. 
1. Notice that the engine image CR state become deploying
1. Verify that Longhorn cannot attach `vol-1` to `node-1` since there is no engine image on `node-1`
1. Check that all functions (attach to other nodes, detach, snapshot, backup, expand, restore, ... ) are working ok for `vol-1`
1. Verify that `vol-2` cannot be attached to any nodes because one of its replica is sitting on the `node-1` which doesn't have the engine image

### Case 3: Test engine upgrade when engine image DaemonSet is not fully deployed
1. Continue from case 2
1. Deploy a new engine image, `newEI`
1. Verify that you can upgrade `vol-1` to `newEI` (both live and offline upgrade)
1. Verify that you can not upgrade `vol-2` to `newEI`

### Case 4: Test replicas scheduling when engine image DaemonSet is not fully deployed
1. Continue from case 2
1. Create a new volume, `vol-3`, with 2 replicas
1. Verify that replicas of `vol-3` are on `node-2` and `node-3`
1. Check that all functions (attach, detach, snapshot, backup, expand, restore, ... ) are working ok for `vol-3`

### Case 5: Test Longhorn upgrade when engine image DaemonSet is not fully deployed
1. Continue from case 3
1. Upgrade Longhorn to a new version
1. Verify that Longhorn successfully upgrade to the new version using the same default engine image even though the default engine image is not fully deployed.

### Case 6: Test `auto upgrade engine` feature when engine image DaemonSet is not fully deployed
Need to re-verify the manual test for feature `auto upgrade engine` https://github.com/longhorn/longhorn-tests/blob/master/docs/content/manual/pre-release/upgrade/auto-upgrade-engine.md
Make sure that Longhorn automatically upgrade engine image for volumes that are either:
1. in the detached state, and has the new engine image on all replicas' node
1. in the attached state, and has the new engine image on all replicas' node and attaching node

**Comment by joshimoo** [comment]
Moving back to implement, since we are missing the deep copy function update.
https://github.com/longhorn/longhorn-manager/blob/75e88c839111b44758c89628d9c99b4148fa0fbc/types/deepcopy.go#L171-L170

**Comment by khushboo-rancher** [comment]
Verified with Longhorn master - `03/18/2021`

Validation - **Fail**

The volume attached to a node where taint was not given, stuck in detaching state on scaling down the pod.
Steps:
1. Create a volume with 3 replicas on a cluster having 3 worker nodes and attach to a pod.
2. Add taint to a node (not attached to the volume)
3. Scale down the pod.
4. Observe the volume stuck in detaching.

[longhorn-support-bundle_482e72ff-5a94-47c1-892c-2e84c290cb23_2021-03-18T11-19-58Z.zip](https://github.com/longhorn/longhorn/files/6163608/longhorn-support-bundle_482e72ff-5a94-47c1-892c-2e84c290cb23_2021-03-18T11-19-58Z.zip)

**Comment by PhanLe1010** [comment]
This problem is tracked at https://github.com/longhorn/longhorn/issues/2081#issuecomment-801842133 since it is not directly related to this feature.

**Comment by yasker** [comment]
Move to Icebox due to blocked by https://github.com/longhorn/longhorn/issues/2377

**Comment by khushboo-rancher** [comment]
Verified with Longhorn v1.1.1-rc1

Validation - **Pass**

The test cases from https://github.com/longhorn/longhorn/issues/2081#issuecomment-783747541 passed.

**Comment by meldafrawi** [comment]
Closing, will track integration test implementation in a separate issue https://github.com/longhorn/longhorn/issues/2480
